File: services/circuit-fine-agent/src/interface/http/routes/directReview.ts
import type { Request, Response } from 'express'
import multer from 'multer'
import path from 'path'
import fs from 'fs'
import { DirectReviewUseCase } from '../../../app/usecases/DirectReviewUseCase.js'
import type { ArtifactStore, ReviewRequest } from '../../../domain/contracts/index.js'

// 中文注释：多文件上传中间件（临时落盘后读取 Buffer，再删除）
export function makeDirectReviewRouter(deps: {
  usecase: DirectReviewUseCase
  artifact: ArtifactStore
  storageRoot: string
}) {
  const uploadDir = path.join(deps.storageRoot, 'tmp')
  try { if (!fs.existsSync(uploadDir)) fs.mkdirSync(uploadDir, { recursive: true }) } catch {}
  const upload = multer({ dest: uploadDir })

  const handler = async (req: Request, res: Response) => {
    try {
      const body = req.body || {}
      const apiUrl = String(body.apiUrl || '')
      const model = String(body.model || '')
      if (!apiUrl || !model) return res.status(400).json({ error: 'apiUrl and model are required' })
      const authHeader = req.header('authorization') || undefined

      // 组装 ReviewRequest
      const filesField = (req as any).files as any[] || []
      const files = filesField.map((f) => {
        const bytes = fs.readFileSync(f.path)
        return { name: f.originalname || f.filename, mime: f.mimetype || 'application/octet-stream', bytes }
      })
      const request: ReviewRequest = {
        files,
        systemPrompt: String(body.systemPrompt || ''),
        requirements: String(body.requirements || ''),
        specs: String(body.specs || ''),
        dialog: String(body.dialog || ''),
        history: (() => { try { return body.history ? (typeof body.history === 'string' ? JSON.parse(body.history) : body.history) : [] } catch { return [] } })(),
        options: { progressId: String(body.progressId || '' ) || undefined }
      }

      const out = await deps.usecase.execute({ apiUrl, model, request, authHeader })
      res.json(out)
    } catch (e: any) {
      res.status(502).json({ error: e?.message || 'upstream error' })
    } finally {
      try {
        const filesField = (req as any).files as any[] || []
        filesField.forEach((f) => { try { fs.unlinkSync(f.path) } catch {} })
      } catch {}
    }
  }

  return { upload, handler }
}

File: services/circuit-agent-py/app/api/routes/direct_review.py
from fastapi import APIRouter, UploadFile, File, Form, Depends, Request
from typing import List, Optional
from app.services.direct_review import DirectReviewUseCase
from app.infra.storage.artifact_store_fs import ArtifactStoreFs
from app.services.timeline_service import TimelineService
from app.infra.providers.vision import OpenRouterVisionProvider
from app.infra.search.openrouter_search import OpenRouterSearch
from app.core.config import get_config
from app.schemas.domain import ReviewRequest, FileItem
import tempfile
import shutil

router = APIRouter()

@router.post('/modes/direct/review')
async def direct_review(request: Request, files: Optional[List[UploadFile]] = File(None)):
    cfg = get_config()
    artifact = ArtifactStoreFs(cfg.storage_root)
    timeline = TimelineService()
    vision = OpenRouterVisionProvider(cfg.openrouter_base, cfg.llm_timeout_ms)
    search = OpenRouterSearch(cfg.openrouter_base, cfg.llm_timeout_ms)
    usecase = DirectReviewUseCase(vision, artifact, timeline, search)

    # parse form fields (要求客户端以表单字段提交 JSON 字段)
    form = await request.form()
    try:
        req_json = form.get('request') or form.get('body') or None
        if req_json:
            import json
            body = json.loads(req_json)
        else:
            # 构建 ReviewRequest from form fields
            body = {}
            for k in ('requirements','specs','dialog','systemPrompt','language'):
                if k in form:
                    body[k] = form.get(k)
        # attach files -> write to temp files
        file_items = []
        if files:
            for f in files:
                suffix = ''
                if f.filename and '.' in f.filename:
                    suffix = '.' + f.filename.split('.')[-1]
                tmp = tempfile.NamedTemporaryFile(delete=False, suffix=suffix)
                try:
                    # stream write
                    while True:
                        chunk = await f.read(1024*64)
                        if not chunk:
                            break
                        tmp.write(chunk)
                    tmp.flush()
                    tmp.close()
                    file_items.append(FileItem(filename=f.filename, mime=f.content_type, path=tmp.name))
                except Exception:
                    try:
                        tmp.close()
                    except Exception:
                        pass
        body['files'] = file_items
        rr = ReviewRequest(**body)
    except Exception as e:
        return { 'error': 'invalid request', 'detail': str(e) }

    auth = request.headers.get('authorization')
    resp = await usecase.execute(api_url=cfg.openrouter_base, model=form.get('model') or 'default', request=rr, auth_header=auth)

    # cleanup temp files
    try:
        for fi in file_items:
            if fi.path:
                try:
                    shutil.os.remove(fi.path)
                except Exception:
                    pass
    except Exception:
        pass

    return resp
